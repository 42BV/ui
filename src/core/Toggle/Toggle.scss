.toggle-container {
  @each $color, $value in $theme-colors {
    &.toggle-#{'' + $color} {
      input[type='checkbox'] {
        &:checked {
          &::after {
            left: 1rem;
            background: darken($value, 10%);
          }

          &::before {
            background: lighten($value, 15%);
          }
        }
      }
    }
  }

  input[type='checkbox'] {
    -webkit-appearance: none;
    appearance: none;
    cursor: pointer;
    position: relative;
    display: inline-block;
    margin-top: 6px;
    margin-right: 4px;
    vertical-align: top;
    width: 34px;

    &:focus {
      outline: none;
    }

    &:checked {
      &::after {
        left: 1rem;
        background: darken(map-get($theme-colors, primary), 10%);
      }

      &::before {
        background: lighten(map-get($theme-colors, primary), 15%);
      }
    }

    &::before,
    &::after {
      display: block;
      margin-right: 4px;
      margin-left: 0;
      content: '';
      position: absolute;
      left: 0;
      @include transition();
    }

    &::after {
      top: -4px;
      border: none;
      width: 1.25rem;
      height: 1.25rem;
      border-radius: 1.25rem;
      background: white;
      box-shadow: 1px 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    }

    &::before {
      width: 34px;
      height: 14px;
      border-radius: 14px;
      border: none;
      background: var(--bs-gray-400);
    }
  }
}
