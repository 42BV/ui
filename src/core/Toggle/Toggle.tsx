import React from 'react';
import classNames from 'classnames';
import { Color } from '../../form/types';

interface ToggleProps {
  /**
   * Optionally the id of the input element
   */
  id?: string;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * The color of the element.
   */
  color: Color;

  /**
   * Optionally the value of the Toggle, when `true` it is active.
   * when `false` it is inactive. Defaults to `false`.
   */
  value?: boolean;

  /**
   * Callback for when the Toggle element is toggled.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional callback for when the Toggle is blurred.
   */
  onBlur?: React.FocusEventHandler;
}

/**
 * Toggle is a component which looks like a switch. Use the Toggle
 * instead of a checkbox, when you want the changes to take effect
 * immediately.
 */
export default function Toggle({
  className,
  color,
  id,
  value = false,
  onChange,
  onBlur
}: ToggleProps) {
  const toggleClasses = classNames(
    'toggle-container',
    `toggle-${color}`,
    className
  );

  return (
    <span className={toggleClasses}>
      <input
        id={id}
        type="checkbox"
        onChange={event => onChange(event.target.checked)}
        checked={value}
        onBlur={onBlur}
      />
    </span>
  );
}
