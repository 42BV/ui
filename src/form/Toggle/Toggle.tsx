import React from 'react';
import classNames from 'classnames';
import { FormGroup, Label } from 'reactstrap';
import { Color } from '../types';

import withJarb from '../withJarb/withJarb';

interface BaseProps {
  /**
   * The color of Toggle element.
   */
  toggleColor: Color;

  /**
   * The value that the form element currently has.
   */
  value?: boolean;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional callback for when the form element is blurred.
   */
  onBlur?: () => void;

  /**
   * Optionally the error message to render.
   */
  error?: React.ReactNode;

  /**
   * Optionally the color of the FormGroup.
   */
  color?: Color;

  /**
   * Whether or not the form element is currently valid.
   */
  valid?: boolean;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
}

interface WithoutLabel extends BaseProps {
  id?: string;
  label?: never;
}

interface WithLabel extends BaseProps {
  /**
   * The id of the form element.
   */
  id: string;

  /**
   * The label of the form element.
   */
  label: string;
}

export type Props = WithoutLabel | WithLabel;

/**
 * FormToggle is a a form element which allows the user to activate or
 * deactivate a certain state.
 */
export function FormToggle(props: Props) {
  const { color, error, className = '' } = props;

  const toggle = getToggle(props);

  return (
    <FormGroup className={className} color={color}>
      {toggle}
      {error}
    </FormGroup>
  );
}

function getToggle(props: Props) {
  const { id, value, toggleColor, onChange, onBlur } = props;

  const toggle = (
    <Toggle
      id={id}
      color={toggleColor}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
    />
  );

  if ('label' in props && props.label) {
    return (
      <Label for={id}>
        {props.label}
        {toggle}
      </Label>
    );
  }

  return toggle;
}

/**
 * Variant of the FormToggle which can be used in a Jarb context.
 */
export const JarbFormToggle = withJarb<boolean, boolean, Props>(FormToggle);

interface ToggleProps {
  /**
   * Optionally the id of the input element
   */
  id?: string;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * The color of the element.
   */
  color: Color;

  /**
   * Optionally the value of the Toggle, when `true` it is active.
   * when `false` it is inactive. Defaults to `false`.
   */
  value?: boolean;

  /**
   * Callback for when the Toggle element is toggled.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional callback for when the Toggle is blurred.
   */
  onBlur?: React.FocusEventHandler;
}

/**
 * Toggle is a component which looks like a switch. Use the Toggle
 * instead of a checkbox, when you want the changes to take effect
 * immediately.
 */
export default function Toggle({
  className,
  color,
  id,
  value = false,
  onChange,
  onBlur
}: ToggleProps) {
  const toggleClasses = classNames(
    'toggle-container',
    `toggle-${color}`,
    className
  );

  return (
    <span className={toggleClasses}>
      <input
        id={id}
        type="checkbox"
        onChange={event => onChange(event.target.checked)}
        checked={value}
        onBlur={onBlur}
      />
    </span>
  );
}
