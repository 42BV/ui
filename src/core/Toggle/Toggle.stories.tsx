import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Toggle from './Toggle';

storiesOf('Core/Toggle', module).add('basic', () => {
  const [primary, setPrimary] = useState(true);
  const [secondary, setSecondary] = useState(true);
  const [info, setInfo] = useState(true);
  const [success, setSuccess] = useState(true);
  const [warning, setWarning] = useState(true);
  const [danger, setDanger] = useState(true);
  const [light, setLight] = useState(true);
  const [dark, setDark] = useState(true);

  return (
    <div className="text-center">
      <Toggle
        value={primary}
        color="primary"
        onChange={setPrimary}
        onBlur={action('onBlur')}
      />
      primary <br />
      <Toggle
        value={secondary}
        color="secondary"
        onChange={setSecondary}
        onBlur={action('onBlur')}
      />
      secondary <br />
      <Toggle
        value={info}
        color="info"
        onChange={setInfo}
        onBlur={action('onBlur')}
      />
      info <br />
      <Toggle
        value={success}
        color="success"
        onChange={setSuccess}
        onBlur={action('onBlur')}
      />
      success <br />
      <Toggle
        value={warning}
        color="warning"
        onChange={setWarning}
        onBlur={action('onBlur')}
      />
      warning <br />
      <Toggle
        value={danger}
        color="danger"
        onChange={setDanger}
        onBlur={action('onBlur')}
      />
      danger <br />
      <Toggle
        value={light}
        color="light"
        onChange={setLight}
        onBlur={action('onBlur')}
      />
      light <br />
      <Toggle
        value={dark}
        color="dark"
        onChange={setDark}
        onBlur={action('onBlur')}
      />
      dark
    </div>
  );
});
