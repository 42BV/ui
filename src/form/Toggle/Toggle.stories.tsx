import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Toggle, { FormToggle, JarbFormToggle } from './Toggle';
import { Form, FinalForm } from '../story-utils';

storiesOf('Form|Toggle', module)
  .add('basic', () => {
    return (
      <div className="text-center">
        <Toggle
          color="primary"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        primary <br />
        <Toggle
          color="secondary"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        secondary <br />
        <Toggle
          color="info"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        info <br />
        <Toggle
          color="success"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        success <br />
        <Toggle
          color="warning"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        warning <br />
        <Toggle
          color="danger"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        danger <br />
        <Toggle
          color="light"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        light <br />
        <Toggle
          color="dark"
          onChange={action('onChange')}
          onBlur={action('onBlur')}
        />
        dark
      </div>
    );
  })
  .add('form', () => {
    return (
      <Form>
        <FormToggle
          id="agree"
          label="Agree"
          toggleColor="primary"
          onChange={value => action(`onChange: ${value}`)}
        />
      </Form>
    );
  })
  .add('without label', () => {
    return (
      <Form>
        <FormToggle
          id="agree"
          toggleColor="primary"
          onChange={value => action(`onChange: ${value}`)}
        />
      </Form>
    );
  })
  .add('jarb', () => {
    return (
      <FinalForm>
        <JarbFormToggle
          id="agree"
          name="agree"
          label="Agree"
          toggleColor="primary"
          jarb={{
            validator: 'Hero.agree',
            label: 'Agree'
          }}
        />
      </FinalForm>
    );
  });
