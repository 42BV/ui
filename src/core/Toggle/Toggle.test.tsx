import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import Toggle from './Toggle';

describe('Component: Toggle', () => {
  let toggle: ShallowWrapper;
  let onChangeSpy: jest.Mock<any, any>;
  let onBlurSpy: jest.Mock<any, any>;

  function setup({ value }: { value?: boolean }) {
    onChangeSpy = jest.fn();
    onBlurSpy = jest.fn();

    toggle = shallow(
      <Toggle
        id="toggle"
        color="primary"
        value={value}
        onChange={onChangeSpy}
        onBlur={onBlurSpy}
      />
    );
  }

  test('ui', () => {
    setup({ value: true });
    expect(toJson(toggle)).toMatchSnapshot('Component: Toggle => ui');
  });

  describe('events', () => {
    test('onChange', () => {
      setup({ value: false });
      const input = toggle.find('input');
      // @ts-expect-error Test mock
      input.props().onChange({ target: { checked: true } });

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(true);

      // @ts-expect-error Test mock
      input.props().onChange({ target: { checked: false } });

      expect(onChangeSpy).toHaveBeenCalledTimes(2);
      expect(onChangeSpy).toHaveBeenCalledWith(false);
    });

    test('onBlur', () => {
      setup({ value: true });
      const input = toggle.find('input');

      // @ts-expect-error Test mock
      input.props().onBlur();
      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('checked changes', () => {
    test('becomes empty', () => {
      setup({ value: false });

      let input = toggle.find('input');
      expect(input.props().checked).toBe(false);

      toggle.setProps({ value: undefined });

      input = toggle.find('input');
      expect(input.props().checked).toBe(false);
    });

    test('becomes filled', () => {
      setup({ value: undefined });

      let input = toggle.find('input');
      expect(input.props().checked).toBe(false);

      toggle.setProps({ value: true });

      input = toggle.find('input');
      expect(input.props().checked).toBe(true);
    });
  });
});
