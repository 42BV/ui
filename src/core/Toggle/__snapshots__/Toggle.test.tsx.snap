// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Toggle ui: Component: Toggle => ui 1`] = `
<span
  className="toggle-container toggle-primary"
>
  <input
    checked={true}
    id="toggle"
    onBlur={[MockFunction]}
    onChange={[Function]}
    type="checkbox"
  />
</span>
`;
